README
HUBO-ACH

Support
=======
Daniel M. Lofaro, dan@danlofaro.com

Run Hubo-Ach
============

1. Setup Config (first time only or new builds and param changes):

    $ hubo-ach config

2. Start hubo-ach daemon.  This will automatically start hubo-console as well
   CAN packets will not start to be sent:

    $ hubo-ach start

3. To stop hubo-ach daemon:

    $ hubo-ach stop

4. To kill all hubo-ach basic processes:

    $ hubo-ach killall

Install ACH
===========
First-time install,
If you have a PEAK CAN Card installed the script below will install
Hubo-Ach and all of it's dependencies for you automatically

Just run:

    $ ./first-time-install-hubo-ach.sh

Or:


1. Add this line to /etc/apt/sources.list:

    deb http://code.golems.org/debian squeeze golems.org

  (Replace squeeze with the codename for your distribution, ie lucid,
  precise, etc.)

2. `sudo apt-get update`
3. `sudo apt-get install libach1 libach-dev`
4. `sudo apt-get install ach-utils`


5) If you have obtained amino from the git repo, you need to
initialize the autotools build scripts.  This requires the autotools
and the autoconf macro archive. To install these on Debian or Ubuntu
GNU/Linux, you can do

    $ sudo apt-get install autoconf automake libtool autoconf-archive

Then, initialize autotools with

    $ autoreconf -i

This whole step is not necessary when you have obtained a distribution
tarball which already contains the autoconf-generated configure script.

Option 1)

    $ ./hubo-ach-install.sh

Option 2)

* Configure for system:

    $ ./configure

* Build:

    $ make

* Install:

    $ sudo make install

Functionality:
==============

* Feedforward:
  - set position reference (rad)

* Feedback:
  - get encoder position (rad)
  - get IMU x angle (rad)
   - get IMU y angle (rad)
   - get IMU Vx velos(rad/sec)
   - get IMU Vy velos(rad/sec)
   - get FT Mx moment (Nm)
   - get FT My moment (Nm)
   - get FT Fz force (N)
   - get foot angle from acc Ax (rad)
   - get foot angle from acc Ay (rad)
   - get foot G-force acc Az (G)
   - get Motor Status (flags)

* Motor Control:
  - home motors
  - reset to defaults

hubo-ach-console
================

Console input for hubo-ach

Usage
-----

Home all motors and zero sensors

    hubo-ach: homeAll
    hubo-ach: iniSensors

Functions
---------

* `initilize XXX` - sets defaults to motor XXX and turns on FETs
* `initilizeAll`  - sets defaults to all motors and turns on all FETs
* `home XXX`      - homes motor XXX
* `homeAll`       - homes all motors that are active
* `goto XXX val`  - sets motor XXX to goto position `val` in rad.
                    uses linear interpolation via hubo-ref-filter
                    0.5 sec linear filter by default
* `get XXX`       - gets current encoder location of motor XXX
* `zero XXX`      - zero the sensor XXX
* `zeroacc`       - zero the foot accelerometer XXX
* `iniSensors`    - zero all sensors
* `ctrl XXX Y`    - turn on and off control to motor XXX.  If Y = 0 then Off, if Y = 1 then On
                    Note: must home joint first.
* `ctrlAll Y`     - turn on and off control to all motors. If Y = 0 then Off, if Y = 1 then On
                    Note: must home joint first.
* `reset XXX`     - Only for motors with error flags raised - no effect on motors with no errors
                    Resets motor XXX encoders and resets all flags.  Also sets reference to 0.0 rad
* `resetAll`      - Only for motors with error flags raised - no effect on motors with no errors
                    Resets all motors and encoder and resets all flags
* `status XXX`    - Gets the most recent status flags for motor XXX

Required for installation
-------------------------
    sudo apt-get install libreadline-dev

For joint "goto" functionality
------------------------------

Must have hubo-ref-filter running

Note: do not have hubo-ref-filter running when trying to write
directly to the hubo-ref channel.

    $ sudo ./hubo-ref-filter


hubo-ach-loop
=============
demo loop that communicates with hubo ach

Install Drivers for PEAK
========================
    sudo apt-get install libpopt-dev
    cd drivers
    tar -zxvf peak-linux-driver-7.7.tar.gz
    cd peak-linux-driver-7.7
    make



hubo-homeparams
===============
For instructions on how to save and load parameters, type
    $ hubo-homeparams
without passing in any arguments, and it will print out this set of instructions:

Usage: hubo-homeparams [argument] [target file] [optional]

arguments
---------

* -s or --save  : Saves the current home settings into [target file]
* -l or --load  : Loads the home settings from [target file] onto the boards
* -p or --print : Prints a table of joint settings out to the terminal
* target file: You are recommend to provide the full path of a file

* optional: Tells the -s/--save argument what format to store position data in
  * raw : Position data will be stored in encoder units
  * rad : Position data will be stored in radians

The important thing is that giving a flag of "-s" will save the
current settings into whatever file you specify, and a flag of "-l"
will load whatever file you specify.

I strongly recommend only using full paths for any target files,
because I have not tested what happens when you give relative paths. I
would guess that it will just fail to find the file.


debian/changelog
hubo-ach (0.1.20131101) stable; urgency=low

  * Added encoder feedback for head process
  * Added head ach channel for encoder feedback
  * Added current feedback for all joints (20x slower then hubo-ach T)
  * Removed duplicate def from include/hubo/canID.h (already in include/hubo.h): HUBO_JOINT_COUNT, HUBO_JMC_COUNT, HUBO_SENSOR_COUNT

 -- Daniel M. Lofaro <dan@danlofaro.com>  Fri, 01 Nov 2013 17:50:00 -0400

hubo-ach (0.1.20131009) stable; urgency=low

  * Added head channel for DRC neck - can command head independent of the rest of the body
  * Added support for RAINBOW neck CAN to Dynamexal controller with encoder feedback
  * Added Dynamexal controller support for neck (baud 57600, Motor 1,2,3)
  * Added power data from back pannel (V, I and P "total")  Note: requires wireing see wiki.danlofaro.com for details
  * Minor bug fixes

 -- Daniel M. Lofaro <dan@danlofaro.com>  Wed, 09 Oct 2013 22:48:00 -0400

hubo-ach (0.1.20130909) stable; urgency=low

  * Logging is disabled by default
  * Joint tables now in hubo-ach-dev

 -- Daniel M. Lofaro <dan@danlofaro.com>  Thurs, 09 Sep 2013 16:00:00 -0400

hubo-ach (0.1.20130808) stable; urgency=low

  * Fixed finger bug in Hubo-Plus model in simulator

 -- Daniel M. Lofaro <dan@danlofaro.com>  Thurs, 08 Aug 2013 17:10:00 -0400

hubo-ach (0.1.20130806) stable; urgency=low

  * Fixed minor simulator bug

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tues, 06 Aug 2013 18:10:00 -0400

hubo-ach (0.1.20130801) stable; urgency=low

  * Changed IMU and tilt sensors to be in rad (previous was in deg)
  * Fixed minor python bug

 -- Daniel M. Lofaro <dan@danlofaro.com>  Thurs, 01 Aug 2013 16:43:00 -0400

hubo-ach (0.1.20130730) stable; urgency=low

  * Added Enc feedback from wrist roll and fingers
  * changed drc model to v3
  * minor fix to PWM duty cycle

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 30 Jul 2013 16:43:00 -0400

hubo-ach (0.1.20130726) stable; urgency=low

  * Added auto install for simulators
  * Added hubo-ach install sim-openhubo
  * Added hubo-ach install sim-drc
  * Added hubo-ach install sim-all
  * Minor bug fixes

 -- Daniel M. Lofaro <dan@danlofaro.com>  Fri, 26 Jul 2013 18:55:00 -0400

hubo-ach (0.1.20130724) stable; urgency=low

  * Minor fix in hubo-ach script dealing with joint table switching

 -- Daniel M. Lofaro <dan@danlofaro.com>  Wed, 24 Jul 2013 14:17:00 -0400

hubo-ach (0.1.20130723-2) stable; urgency=low

  * Fixed bug in hubo-ach script for joint table changing
  * Added IMU and FT name in hubo-read

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 23 Jul 2013 14:35:00 -0400

hubo-ach (0.1.20130723) stable; urgency=low

  * Fixed hubo_ach.py bug
  * Fixed OpenHubo bug

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 23 Jul 2013 13:31:00 -0400

hubo-ach (0.1.20130722) stable; urgency=low

  * Auto Logging with limiting filesize
  * Switch mode CAN frame added for hand boards

 -- Daniel M. Lofaro <dan@danlofaro.com>  Mon, 22 Jul 2013 17:55:00 -0400

hubo-ach (0.1.20130719) stable; urgency=low

  * Added auto state logging

 -- Daniel M. Lofaro <dan@danlofaro.com>  Fri, 19 Jul 2013 18:06:00 -0400

hubo-ach (0.1.20130716) stable; urgency=low

  * Added compliance mode
  * added joint home default positions

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 16 Jul 2013 18:00:00 -0400

hubo-ach (0.1.20130702-2) stable; urgency=low

  * Added Sensor Joint Tables

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 02 Jul 2013 15:38:00 -0400

hubo-ach (0.1.20130702) stable; urgency=low

  * DRC Hubo hands working
  * DRC Hubo Wrist Roll working
  * Changed direction of FT
  * Added Pip and ach python install on update

 -- Daniel M. Lofaro <dan@danlofaro.com>  Tue, 02 Jul 2013 12:56:00 -0400

hubo-ach (0.1.20130628) stable; urgency=low

  * disabled hands on joint table

 -- Daniel M. Lofaro <dan@danlofaro.com>  Fri, 28 Jun 2013 16:45:00 -0400

hubo-ach (0.1.20130626) stable; urgency=low

  * added drc-hubo joint table

 -- Daniel M. Lofaro <dan@danlofaro.com>  Mon, 24 Jun 2013 12:51:00 -0400

hubo-ach (0.1.20130624) stable; urgency=low

  * added drc-hubo joint table

 -- Daniel M. Lofaro <dan@danlofaro.com>  Mon, 24 Jun 2013 12:51:00 -0400

hubo-ach (0.1.20130520) stable; urgency=low

  * Fixed problemm with the FT sensors, all FT reads now, just initilize
  * initilize with iniSensors in hubo-console

 -- Daniel M. Lofaro <dan@danlofaro.com>  Mon, 20 May 2013 12:26:00 -0400

hubo-ach (0.1.20130515) stable; urgency=low

  * added joint names to hubo.h and removed them from jointparams.c

 -- Daniel M. Lofaro <dan@danlofaro.com>  Wed, 15 May 2013 16:09:00 -0400

hubo-ach (0.1.20130420-2) stable; urgency=low

  * Added statusAll to to hubo-console

 -- Daniel M. Lofaro <dan@danlofaro.com>  Sat, 20 Apr 2013 21:50:00 -0400

hubo-ach (0.1.20130420) stable; urgency=low

  * OpenHubo No Physics on Remote working

 -- Daniel M. Lofaro <dan@danlofaro.com>  Sat, 20 Apr 2013 12:32:00 -0400

hubo-ach (0.1.20130411) stable; urgency=low

  * Left hand working again

 -- Daniel M. Lofaro <dan@danlofaro.com>  Thu, 11 Apr 2013 16:30:00 -0400

hubo-ach (0.1.20130408) stable; urgency=low

  * fixed encoder read back in physics mode

 -- Daniel M. Lofaro <dan@danlofaro.com>  Mon, 08 Apr 2013 16:38:00 -0400

hubo-ach (0.1.20130329) stable; urgency=low

  * sim dynamic and nodynamics to physics and nophysics

 -- Daniel M. Lofaro <dan@danlofaro.com>  Fri, 29 Mar 2013 01:14:00 -0400

hubo-ach (0.1.20130327) stable; urgency=low

  * Fixed small issue with RAP in hubo_ach.py

 -- Daniel M. Lofaro <dan@danlofaro.com>  Wed, 27 Mar 2013 15:15:00 -0400

hubo-ach (0.0.20130326) stable; urgency=low

  * added hubo-ach update (updates from dpkg from website lates release)
  * added hubo-ach update develop (updates from dpkg from website latest develop)
  * added hubo-ach update apt-get (updates from apt-get what ever repo is on /etc/apt/soruce.list)

hubo.h
/* -*-	indent-tabs-mode:t; tab-width: 8; c-basic-offset: 8  -*- */
#include "canId.h"
#include <stdint.h>

//#define true 1;
//#define false 0;
//888888888888888888888888888888888888888888
//-----[Static Definitions and Offsets]-----
//888888888888888888888888888888888888888888


/* Tx Message ID */

// Motor Command Message ID
#define txMoCmdId		0x01

// Sensor Command Message ID
#define txSeCmdId		0x02

// Reference Message ID 	(txReMsID + BNO)
#define txReMesId		0x10



/* Rx Message ID */

// FT sensor data Message ID 		(rxFtMsgId + SBNO)
#define rxFtMsgId		0x40

// IMU and Tilt Sensor data Message ID 	(rxImuMsgId + SBNO)
#define rxImuMsgId		0x50

// Encoder Value Message ID 		(rxEncValId + BNO)
#define rxEncValId 		0x60

// Status Message ID 			(rxStaMsgId + BNO)
#define rxStaMsgId		0x150

// Board Information Message ID 	(rxBrdInfoId + BNO + BOFF)
#define rxBrdInfoId		0x190

// Board para and Current Message ID 	(rxBrdCurMsgId + BNO)
#define rxBrdCurMsgId 		0x1C0

// BNO = Board number
// SBNO = sensor board number SBNO=BNO-0x2F
// BOFF=0 	for BNO < 0x30
// BOFF=0x80 	for BNO >= 0x30


/* Joint Numbers/Index values */
#define		RHY		26		//	Right Hip Yaw
#define		RHR		27		//	Right Hip Roll
#define		RHP		28		//	Right Hip Pitch
#define		RKN		29		//	Right Knee Pitch
#define		RAP		30		//	Right Ankle Pitch
#define		RAR		31		//	Right Ankle Roll

#define		LHY		19		//	Left Hip Yaw
#define		LHR		20		//	Left Hip Roll
#define		LHP		21		//	Left Hip Pitch
#define		LKN		22		//	Left Knee Pitch
#define		LAP		23		//	Left Ankle Pitch
#define		LAR		24		//	Left Ankle Roll

#define		RSP		11		//	Right Shoulder Pitch
#define		RSR		12		//	Right Shoulder Pitch
#define		RSY		13		//	Right Shoulder Roll
#define		REB		14		//	Right Elbow Pitch
#define		RWY		15		// right wrist yaw
#define		RWR		16		// right wrist roll
#define		RWP		17		// right wrist Pitch

#define		LSP		3		//	Left Shoulder Pitch
#define		LSR		4		//	Left Shoulder Yaw
#define		LSY		5		//	Left Shoulder Roll
#define		LEB		6		//	Left Elbow Pitch
#define		LWY		7		// left wrist yaw
#define		LWR		8		// left wrist roll
#define		LWP		9		// left wrist pitch

#define		NKY		1		// neck yaw
#define		NKP		2		// neck pitch
#define		NK1		2		// neck 1
#define		NK2		2		// neck 2

#define		WST		0		//	Trunk Yaw

#define		RF1		32		//	Right Finger
#define		RF2		33		//	Right Finger
#define		RF3		34		//	Right Finger
#define		RF4		35		//	Right Finger
#define		RF5		36		//	Right Finger
#define		LF1		37		//	Left Finger
#define		LF2		38		//	Left Finger
#define		LF3		39		//	Left Finger
#define		LF4		40		//	Left Finger
#define		LF5		41		//	Left Finger


#define		HUBO_JOINT_COUNT	50		// 	the size of the array
						//	for the joints
#define 	HUBO_JMC_COUNT		0X40	// 	numbher of jmc
//#define		numOfCmd	3		//  	number of commiands
//#define 	numOfJmc	0x40		//	number of JMCs
#define 	pi		3.141596

#define		HUBO_CHAN_REF_NAME       "hubo-ref"        // hubo ach channel
#define		HUBO_CHAN_INIT_CMD_NAME	 "hubo-init-cmd"   // hubo console channel for ach
#define		HUBO_CHAN_STATE_NAME     "hubo-state"      // hubo state ach channel
#define		HUBO_CHAN_PARAM_NAME     "hubo-param"      // hubo param ach channel


/* def for console do flags */
/* unless otherwise noted cmd[0] = command, cmd[1] = motor# */
typedef enum {
	HUBO_JMC_INI 		= 1,	///< Initilize jmc
	HUBO_FET_ON_OFF 	= 2,	///< turn fet on or off cmd[2] = 1 (on), 0 (off)
	HUBO_CTRL_ON_OFF 	= 3,	///< turn control on or off cmd[2] = 1 (on), 0 (off)
	HUBO_ZERO_ENC		= 4,	///< zero encoder for given motor
	HUBO_GOTO_REF		= 5,	///< go to ref val[0] = ref (rad)
	HUBO_JMC_BEEP		= 6,	///< make beep val[0] = beep time in sec
	HUBO_GOTO_HOME		= 7	///< go home position
} hubo_console_t;

typedef enum {
	HUBO_FT_R_HAND    = 0, ///< Index of right hand FT
	HUBO_FT_L_HAND    = 1, ///< Index of left hand FT
	HUBO_FT_R_FOOT    = 2, ///< Index of right foot FT
	HUBO_FT_L_FOOT    = 3  ///< Index of left foot FT
} hubo_ft_index_t;



struct hubo_joint_param {
	uint16_t motNo;		///< joint number (on board i.e. 0, 1, 2)
	uint16_t jntNo;		///< what overall number joint is it i.e. what RSP=23
	uint32_t refEnc; 	///< encoder reference
	uint16_t drive;		///< size of drive wheel
	uint16_t driven;	///< size of driven wheel
	uint16_t harmonic;	///< gear ratio of harmonic drive
	uint16_t enc;		///< encoder size
	uint8_t dir;		///< direction
	char name[4];		///< name
	uint16_t jmc;		///< motor controller number
	uint8_t can;		///< can channel
	uint8_t active; 	///< checks if the joint is active or not
	uint8_t numMot;		///< number of motors
	uint8_t zeroed;		///< checks to see if the motor is zeroed
};
//}__attribute__((packed));

struct hubo_joint_state {
	double pos;     ///< actual position (rad)
	double cur;     ///< actual current (amps)
	double vel;     ///< actual velocity (rad/sec)
};

struct hubo_ft {
	double m_x;	///< Moment in X (Mx)
	double m_y;       ///< Moment in Y (My)
	double f_z;       ///< Force in Z (Fz)
};

struct hubo_imu {
	double w_x;    ///< rotational velocity in x (rad/s)
	double w_y;    ///< rotational velocity in y (rad/s)
	double w_z;    ///< rotational velocity in z (rad/s)
	double a_x;    ///< linear acceleration in x (m/s/s)
	double a_y;    ///< linear acceleration in y (m/s/s)
	double a_z;    ///< linear acceleration in z (m/s/s)
};

struct hubo_ref {
	double ref[HUBO_JOINT_COUNT];	///< joint reference
	struct timespec time;           ///< time message sent
};

struct hubo_state {
	struct hubo_imu imu;	///< IMU
	struct hubo_ft ft[4];   ///< ft sensors
};

struct hubo_init_cmd {
	/* values for console commands */
	double val[3];
	uint16_t cmd[3];
};

struct jmcDriver{
	uint8_t jmc[5]; // other motors on the same drive
};

struct hubo_param {
	struct hubo_joint_param joint[HUBO_JOINT_COUNT];
	struct jmcDriver driver[HUBO_JMC_COUNT];	// motor driver conneciton info
};

extern int hubo_debug;
